vicodyn/src/vicodyn/proxy/endpoint.cpp

```diff
@@ -0,0 +1,149 @@
+#include "cocaine/vicodyn/proxy/endpoint.hpp"
+
+#include "cocaine/repository/vicodyn/balancer.hpp"
+
+#include <blackhole/wrapper.hpp>
+#include <cocaine/format.hpp>
+#include <cocaine/format/exception.hpp>
+
+namespace cocaine {
+namespace vicodyn {
+
+endpoint_t::endpoint_t(const std::string& dispatch_name, const std::string& app_name, std::size_t buffer_capacity,
+                api::vicodyn::balancer_ptr balancer, peers_t& peers, std::shared_ptr<request_context_t> request_context,
+                logging::logger_t& logger)
+    : buffer_(buffer_capacity)
+    , backward_dispatch_(dispatch_name)
+    , app_name_(app_name)
+    , peers_(peers)
+    , balancer_(std::move(balancer))
+    , request_context_(std::move(request_context))
+    , base_logger_(logger) {
+}
+
+auto endpoint_t::enqueue(const hpack::headers_t& headers, std::string&& event,
+                std::weak_ptr<vicodyn_dispatch_t> life_cycle) -> void {
+    life_cycle_ = std::move(life_cycle);
+    buffer_.store_event(headers, event);
+    if (!choose_endpoint(headers, std::move(event))) {
```

Comment:
If `choose_endpoint()` will be failed here, we will not reconnect to chosen peer.